src/coreclr/jit/lowerarmarch.cpp

```diff
@@ -2266,156 +2270,105 @@ void Lowering::ContainCheckCompare(GenTreeOp* cmp)
     {
-        // Count both sides.
-        return IsValidCompareChain(child->AsOp()->gtGetOp2(), child) &&
-               IsValidCompareChain(child->AsOp()->gtGetOp1(), child);
+        return nullptr;
     }
-    else if (child->OperIsCmpCompare() && varTypeIsIntegral(child->gtGetOp1()) && varTypeIsIntegral(child->gtGetOp2()))
+
+    if (!op2->OperIsCmpCompare() || !varTypeIsIntegral(op2->gtGetOp1()))
     {
-        // Can the child compare be contained.
-        return IsInvariantInRange(child, parent);
+        return nullptr;
     }
 
-    return false;
-}
-
-//------------------------------------------------------------------------
-// ContainCheckCompareChain : Determine if a chain of ANDs and CMPs can be contained.
-//
-// Arguments:
-//    child - pointer to the node being checked.
-//    parent - parent node of the child.
-//    startOfChain - If found, returns the earliest valid op in the chain.
-//
-// Return value:
-//    True if a valid chain is was contained.
-//
-// Notes:
-//    Assumes the chain was checked via IsValidCompareChain.
-//
-bool Lowering::ContainCheckCompareChain(GenTree* child, GenTree* parent, GenTree** startOfChain)
-{
-    assert(parent->OperIs(GT_AND) || parent->OperIs(GT_SELECT));
-    *startOfChain = nullptr; // Nothing found yet.
-
-    if (parent->isContainedCompareChainSegment(child))
+    // We leave checking invariance of op1 to tree to TryLowerConditionToFlagsNode.
+    if (!IsInvariantInRange(op2, tree))
     {
-        // Already have a contained chain.
-        return true;
+        return nullptr;
     }
-    // Can the child be contained.
-    else if (IsInvariantInRange(child, parent))
+
+    GenCondition cond1;
+    if (!TryLowerConditionToFlagsNode(tree, op1, &cond1))
     {
-        if (child->OperIs(GT_AND))
-        {
-            // If Op2 is not contained, then try to contain it.
-            if (!child->isContainedCompareChainSegment(child->AsOp()->gtGetOp2()))
-            {
-                if (!ContainCheckCompareChain(child->gtGetOp2(), child, startOfChain))
-                {
-                    // Op2 must be contained in order to contain Op1 or the AND.
-                    return false;
-                }
-            }
+        return nullptr;
+    }
 
-            // If Op1 is not contained, then try to contain it.
-            if (!child->isContainedCompareChainSegment(child->AsOp()->gtGetOp1()))
-            {
-                if (!ContainCheckCompareChain(child->gtGetOp1(), child, startOfChain))
-                {
-                    return false;
-                }
-            }
+    BlockRange().Remove(op2);
+    BlockRange().InsertBefore(tree, op2);
 
-            // Contain the AND.
-            child->SetContained();
-            return true;
-        }
-        else if (child->OperIsCmpCompare())
-        {
-            child->AsOp()->SetContained();
+    GenCondition cond2 = GenCondition::FromRelop(op2);
+    op2->SetOper(GT_CCMP);
```

Comment:
(I know this is done elsewhere), but why is this safe? op2 is moving from a GenTreeOp to a GenTreeCCMP. The latter class is bigger than the former. What's stopping memory overflow?

Reply:
`GenTree::SetOper` also has asserts that the bashing done is safe as far as the size issue goes.

Reply:
Right, hoped there was something like that. 
Looking at the static assert checks for TREE_NODE_SZ_SMALL / TREE_NODE_SZ_LARGE, there's nothing there for GenTreeConditional

Reply:
Added one for `GenTreeConditional` with the latest commits